types/src/executable.rs

```diff
@@ -18,17 +15,18 @@ pub enum ExecutableDescriptor {
 }
 
 pub trait ModulePath {
-    fn module_path(&self) -> Option<AccessPath>;
+    fn is_module_path(&self) -> bool;
```

Comment:
I find this confusing, but understand the confusion does not root from your PR.

IIUC the relationship here can be more precisely described as "<type> has a variant that is a module path". Not sure how easy it is to model this though..

Reply:
yes, ideally down here it'd already be debug or occasional asserts, and controlled on the highest caller side.

Reply:
it's not particularly cheap and still must be happening more times than it should be possible with cleaner APIs and distinction, if I had to guess.

Reply:
@georgemitenkov looks like we still want to make some more improvements to this in the future. Can you add a TODO here so that we don't forget?

Reply:
done